Add bubble and insertion sort algorithms

patrickmarabeas opened the pull request:
Implements the bubble and insertion sort algorithms. Both functions take a slice of n ints and return a sorted slice of n ints using specified sorting algorithm.

- Adds bubble sort algorithm
- Adds insertion sort algorithm

Fixes #417 

Review of colleague's PR #501

Marchitect commented:
Looks all good to me, I think your solutions look neat, good work. One suggestion I can give would be that you might consider adding negative integers into test cases.

patrickmarabeas commented:
@Marchitect - Added negative integer tests.

undrewb reviewed (dismissed):
Nice neat implementation. Plenty of test cases. Good work!

anzdaddy commented on 02_bubble/patrickmarabeas/main.go:
```diff
@@ -0,0 +1,59 @@
+	length := len(s)
+	// Create copy
+	result := make([]int, length)
+	copy(result, s)
```
This forces a copy that the caller may not require. Making a copy should be left up to the caller.

patrickmarabeas replied:
This is a good point, thanks. I'll make sure I'm not being opinionated re immutability in the future.

anzdaddy commented on 02_bubble/patrickmarabeas/main.go:
```diff
@@ -0,0 +1,53 @@
+var out io.Writer = os.Stdout
+
+// Bubble implements the Bubble sort algorithm.
+func bubble(s []int) []int {
```
nit: With the change in mutability behaviour, it would be clearer to not return anything.

patrickmarabeas replied:
~Would this result in the following signature `func bubble(s *[]int)`?~

I see Slices are passed by reference already.

patrickmarabeas replied:
Added alternative implementation and tests

anzdaddy commented on 02_bubble/patrickmarabeas/main_test.go:
```diff
@@ -0,0 +1,92 @@
+func TestBubble2(t *testing.T) {
+	for name, c := range cases {
+		got := c.input
+		bubble2(got)
```
Don't forget that you're now mutating the original and thus might require a copy.

patrickmarabeas replied:
Reverted to returning a copy, as per spec

juliaogris reviewed (commented):
Please address [failing build]().  
From the [lab assignment doc](https://github.com/anz-bank/go-course/blob/master/02_bubble/README.md):
> Write a function that returns a sorted *copy* 
(my emphasis).  

You are mutating the input.

patrickmarabeas committed:
Add bubble sort algorithm

Implements a function which takes a slice of n ints and returns a
sorted slice of n ints using the bubble sort algorithm

Lab 02 (#417)

patrickmarabeas committed:
Add insertion sort algorithm

Implements a function which takes a slice of n ints and returns a
sorted slice of n ints using the insertion sort algorithm

Lab 02 (#417)

patrickmarabeas commented:
Reverted to original implementation

juliaogris reviewed (approved):
LGTM.

Just a minor nit on style, don't comment the obvious, that just creates noise:

```go
	// Create copy
	result := make([]int, length)
	copy(result, s)
```

patrickmarabeas committed:
Merge branch 'master' into lab-02-pm
